File: benchmarks/tsb/bench_multi_index_droplevel.ts
/**
 * Benchmark: MultiIndex droplevel, reorderLevels, and setNames
 */
import { MultiIndex } from "../../src/index.js";

const ROWS = 100_000;
const WARMUP = 3;
const ITERATIONS = 10;

const a = Array.from({ length: ROWS }, (_, i) => `a${i % 100}`);
const b = Array.from({ length: ROWS }, (_, i) => i % 1000);
const c = Array.from({ length: ROWS }, (_, i) => i % 50);
const tuples: [string, number, number][] = a.map((v, i) => [v, b[i], c[i]]);
const mi = new MultiIndex({ tuples, names: ["x", "y", "z"] });

for (let i = 0; i < WARMUP; i++) {
  mi.droplevel(0);
  mi.reorderLevels([2, 1, 0]);
  mi.setNames(["a", "b", "c"]);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  mi.droplevel(0);
  mi.reorderLevels([2, 1, 0]);
  mi.setNames(["a", "b", "c"]);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "multi_index_droplevel",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_multi_index_droplevel.py
"""Benchmark: MultiIndex droplevel, reorder_levels, set_names"""
import json, time
import pandas as pd

ROWS = 100_000
WARMUP = 3
ITERATIONS = 10

a = [f"a{i % 100}" for i in range(ROWS)]
b = [i % 1000 for i in range(ROWS)]
c = [i % 50 for i in range(ROWS)]
tuples = list(zip(a, b, c))
mi = pd.MultiIndex.from_tuples(tuples, names=["x", "y", "z"])

for _ in range(WARMUP):
    mi.droplevel(0)
    mi.reorder_levels([2, 1, 0])
    mi.set_names(["a", "b", "c"])

start = time.perf_counter()
for _ in range(ITERATIONS):
    mi.droplevel(0)
    mi.reorder_levels([2, 1, 0])
    mi.set_names(["a", "b", "c"])
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "multi_index_droplevel",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
